File: cookbooks/typescript/hn-time-capsule/pipeline.ts
/**
 * HN Time Capsule - Evolve Edition
 *
 * Karpathy's 1,486 lines -> ~50 lines
 */

import "dotenv/config";
import { writeFileSync, mkdirSync, rmSync } from "fs";
import { Swarm, Pipeline } from "@evolvingmachines/sdk";
import { FETCH, ANALYZE, RENDER } from "./prompts";
import { AnalysisSchema } from "./schema";
import { saveIntermediate } from "./utils";

const swarm = new Swarm({
    tag: "hn-time-capsule",
    concurrency: 10,
    retry: { maxAttempts: 2 },
});

const pipeline = new Pipeline(swarm)
    .map({
        name: "fetch",
        prompt: FETCH,
        agent: { type: "claude", model: "haiku" },
    })
    .map({
        name: "analyze",
        prompt: ANALYZE,
        schema: AnalysisSchema,
        agent: { type: "claude", model: "haiku" },
    })
    .reduce({
        name: "render",
        prompt: RENDER,
    });

async function main() {
    // Clean previous run
    rmSync("intermediate", { recursive: true, force: true });
    rmSync("output", { recursive: true, force: true });

    const date = "2015-12-01";
    const limit = 30;

    const items = Array.from({ length: limit }, (_, i) => ({
        "config.json": JSON.stringify({ rank: i + 1, date }),
    }));

    console.log(`Processing ${items.length} articles from ${date}...`);
    const result = await pipeline.run(items);

    saveIntermediate(result.steps[0].results as any[], "fetch");
    saveIntermediate(result.steps[1].results as any[], "analyze");

    mkdirSync("output", { recursive: true });
    for (const [name, content] of Object.entries(result.output.files)) {
        writeFileSync(`output/${name}`, content as string);
    }
    console.log("Done! Output saved to ./output/");
}

main().catch(console.error);

File: cookbooks/python/hn-time-capsule/pipeline.py
"""
HN Time Capsule - Evolve Edition

Karpathy's 1,486 lines → 50 lines
"""

import json
import shutil
from pathlib import Path
from dotenv import load_dotenv
from evolve import Swarm, SwarmConfig, Pipeline, MapConfig, ReduceConfig, RetryConfig, AgentConfig
from prompts import FETCH, ANALYZE, RENDER
from schema import Analysis
from utils import save_intermediate

load_dotenv()


swarm = Swarm(SwarmConfig(
    tag='hn-time-capsule',
    concurrency=1000,
    retry=RetryConfig(max_attempts=2),
))

pipeline = (
    Pipeline(swarm)
    .map(MapConfig(
        name='fetch',
        prompt=FETCH,
        agent=AgentConfig(type='claude', model='haiku'),
    ))
    .map(MapConfig(
        name='analyze',
        prompt=ANALYZE,
        schema=Analysis,
        agent=AgentConfig(type='claude', model='haiku'),
    ))
    .reduce(ReduceConfig(
        name='render',
        prompt=RENDER,
    ))
)


async def main():
    # Clean previous run
    shutil.rmtree("intermediate", ignore_errors=True)
    shutil.rmtree("output", ignore_errors=True)

    date = "2015-12-01"
    limit = 30

    items = [{"config.json": json.dumps({"rank": i, "date": date})} for i in range(1, limit + 1)]

    print(f"Processing {len(items)} articles from {date}...")
    result = await pipeline.run(items)

    save_intermediate(result.steps[0].results, "fetch")
    save_intermediate(result.steps[1].results, "analyze")

    Path("output").mkdir(exist_ok=True)
    for name, content in result.output.files.items():
        Path(f"output/{name}").write_text(content)
    print(f"Done! Output saved to ./output/")


if __name__ == "__main__":
    import asyncio
    asyncio.run(main())
